src/tester.rs

```diff
@@ -37,52 +38,128 @@ pub fn run_test(run: &mut Command, testcase: &ClashTestCase) -> Result<TestRunRe
         }
 
         TestRunResult::WrongOutput { stderr, stdout } => {
-            println!("{} {}", failure_style.paint("FAIL"), title);
+            println!("{} {}", style.failure.paint("FAIL"), title);
             if !stderr.is_empty() {
-                println!("{}", stderr_style.paint(stderr.trim_end()));
+                println!("{}", style.stderr.paint(stderr.trim_end()));
             }
-            let formatter = Formatter::default();
-            println!(
-                "==== EXPECTED ====\n{}",
-                formatter.show_whitespace(&testcase.test_out, &out_style, &ws_style)
-            );
-            println!(
-                "===== ACTUAL =====\n{}\n",
-                formatter.show_whitespace(stdout, &out_style, &ws_style)
-            );
+
+            let input = &testcase.test_in;
+            let expected = &testcase.test_out;
+            let received = &stdout;
+            // compare in bulk
+            print_pair(input, expected, received, &formatter, &style);
+            println!("");
+            // compare but stop printing RECEIVED when an error is found (+format)
+            compare(input, expected, received, &formatter, &style);
         }
 
-        TestRunResult::RuntimeError { stderr } => {
-            println!("{} {}", error_style.paint("ERROR"), title);
-            println!("{}\n", stderr_style.paint(stderr.trim_end()));
+        TestRunResult::RuntimeError { stdout, stderr } => {
+            println!("{} {}", style.error.paint("ERROR"), title);
+            println!("{}\n", style.stderr.paint(stderr.trim_end()));
+            print_pair(&testcase.test_in,&testcase.test_out, &stdout, &formatter, &style);
         }
     }
 }
+
+fn compare(input: &str, expected: &str, received: &str, formatter: &Formatter, style: &TestCaseStyle) {
+    let errors_allowed = 10;
+
+    // If nothing was received at all, special message?
+    if received.is_empty() {
+        print_pair(input, expected, &Color::Red.paint("Nothing").to_string(), &formatter, style);
+        return;
+    }
+
+    let mut buffer = String::new();
+    let mut error_count = 0;
+    let mut iexp = 0;
+    let mut irec = 0;
+
+    while iexp < expected.chars().count() && irec < received.chars().count() {
+        let cexp = expected.chars().nth(iexp).unwrap();
+        let crec = received.chars().nth(irec).unwrap();
+    
+        if crec == cexp {
+            buffer += &formatter.show_whitespace(&crec.to_string(), &style.out, &style.whitespace);
+        } else {
+            buffer += &style.failure.paint(crec.to_string()).to_string();
+            error_count += 1;
+        }
+        // To many errors, stop here
+        if error_count > errors_allowed {
+            let fmt_received = format!("{}{}", buffer, &style.failure.paint("..."));
+            print_pair(input, expected, &fmt_received, &formatter, &style);
+            return;
+        }
+        iexp += 1;
+        irec += 1;
+    }
```

Comment:
Isn't this whole loop just a [`.zip()`](https://doc.rust-lang.org/std/iter/trait.Iterator.html#method.zip)? I don't like keeping track of index variables when it can be avoided (also less need for `.unwrap()`).